scripts/artifacts.py

```diff
@@ -0,0 +1,263 @@
+                          }
+                        }
+                      }
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  """
+
+    vars = {"ref": ref, "limit": limit}
+
+    ret = requests.post(
+        ENDPOINT_GRAPHQL, json={"query": query, "variables": vars}, headers=headers
+    )
+    ret.raise_for_status()
+    ret = ret.json()
+
+    commit_infos = [
+        CommitInfo(
+            message=get_first_line(commit["message"]),
+            ref=ref if i == 0 else f"{ref}~{i}",
+            rev=commit["oid"],
+            workflow_ids=[
+                check["node"]["checkSuite"]["workflowRun"]["databaseId"]
+                for check in commit["statusCheckRollup"]["contexts"]["edges"]
+            ]
+            if commit["statusCheckRollup"]
+            else [],
+        )
+        for i, commit in enumerate(
+            ret["data"]["repository"]["object"]["history"]["nodes"]
+        )
+        if commit["statusCheckRollup"]
+    ]
+
+    return commit_infos
+
+
+def fetch_commit_info(expr):
+    return fetch_commit_infos(expr, limit=1)[0]
+
+
+def mk_commit_stats(info):
+    stats = download_stats(info.workflow_ids)
+
+    gas_costs = stats.get("gas-costs", [{}])[0]
+    # TODO: Gas costs are strings for no reason.
+    gas_costs = {k: int(v) for k, v in gas_costs.items()}
+
+    entrypoint_sizes = stats.get("entrypoint-sizes", [{}])[0]
+
+    return CommitStats(
+        info=info, gas_costs=gas_costs, entrypoint_sizes=entrypoint_sizes
+    )
+
+
+# Given a list of workflow_ids, download the JSON files within the artifacts.
+def download_stats(workflow_ids):
+    stats_urls = []
+
+    # fetch all artifacts archive urls
+    for workflow_id in workflow_ids:
+        url = (
+            ENDPOINT_REST
+            + f"/repos/tezos-checker/checker/actions/runs/{workflow_id}/artifacts"
+        )
+        ret = requests.get(url, headers=headers)
+        ret.raise_for_status()
+        ret = ret.json()
+        stats_urls.extend(i["archive_download_url"] for i in ret["artifacts"])
+
+    stat_results = {}
+    for url in stats_urls:
+        req = requests.get(url, headers=headers, stream=True)
+        buf = io.BytesIO(req.content)
+        with zipfile.ZipFile(buf) as zip:
+            for info in zip.infolist():
+                name, ext = os.path.splitext(info.filename)
+                if ext == ".json":
+                    ret = json.loads(zip.read(info).decode("utf-8"))
+                    stat_results.setdefault(name, []).append(ret)
+
+    return stat_results
+
+
+def get_first_line(s):
+    ls = s.splitlines()
+    return ls[0] if ls else ""
+
+
+def short_rev(rev):
+    return rev[:8]
+
+
+def list_getdefault(l, ix, d):
+    return l[ix] if ix < len(l) else d
+
+
+def diff_change(diff):
+    if diff.previous is None:
+        return diff.next
+    elif diff.next is None:
+        return diff.previous
```

Comment:
Nitpick: I guess for consistency this should be negative:
```suggestion
        return (- diff.previous)
```
I wonder also what would happen if both were `None`. Perhaps we could first check for that:
```python
    if diff.previous is None and diff.next is None:
        return 0
    elif diff.previous is None:
        ...
```

Reply:
They both can not be `None` (unfortunately, not by construction); since having `None` on one side means that the key is found in one direction but not on the other. When the key is not there at all, we won't have the `Diff` object. If this was Haskell, we would be using the `these` package :).

> Nitpick: I guess for consistency this should be negative:

That's right, thanks!

Reply:
Ah, I see. Fair enough

> If this was Haskell, we would be using the these package :).

:heart: